xds/src/main/java/io/grpc/xds/sds/TlsCertificateSecretVolumeSecretProvider.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2019 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds.sds;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.protobuf.ByteString;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ *  Implementation of a file based secret provider.
+ */
+final class TlsCertificateSecretVolumeSecretProvider
+    implements SecretProvider<TlsCertificateStore> {
+
+  private static final Logger logger = Logger
+      .getLogger(TlsCertificateSecretVolumeSecretProvider.class.getName());
+  public static final String PEM = ".pem";
+  public static final String CRT = ".crt";
+
+  private final String path;
+
+  // for now mark it unused
+  @SuppressWarnings("unused")
+  private final String name;
+
+  TlsCertificateSecretVolumeSecretProvider(String path, String name) {
+    this.path = path;
+    this.name = name;
+  }
+
+  @Override
+  public void addListener(Runnable listener, Executor executor) {
```

Comment:
i don't understand this method. is the listener supposed be called when the future (or `get()`) is finished? i know this class is more like immediate future. but how the user use this listener?

let's use the generic async case since this specific one is not generic case.

I imagine when the `get()` has result / failed, it will trigger the listeners (Runnables). Since it is vanilla `Runnable`, the listener has no access to the value of the result, so the runnable must call the `get()` again to access the result? (will this trigger the new refresh?) or the listener doesn't need to use the result?

are those listeners will be triggered everytime the secret is re-evaluated? how do we know when to refresh vs use cache?

let say i only need the first update, how to opt out this notification?

Reply:
based on offline discussion of the design, I am hoping this is resolved